snorkel/contrib/lstm/lstm_base.py

```diff
@@ -0,0 +1,209 @@
+import numpy as np
+import random
+import tensorflow as tf
+import tensorflow.contrib.rnn as rnn
+
+from snorkel.learning import LabelBalancer, TFNoiseAwareModel
+from utils import f1_score, get_bi_rnn_output, SymbolTable
+from time import time
+
+
+SD = 0.1
+
+
+class LSTMBase(TFNoiseAwareModel):
+
+    representation = True
+
+    def __init__(self, save_file=None, name='LSTMBase', seed=None, n_threads=4):
+        """Base class for LSTM"""
+        # Define metadata
+        self.mx_len    = None # Max sentence length
+        self.dim       = None # Embedding dimension
+        self.n_v       = None # Vocabulary size
+        self.lr        = None # Learning rate
+        self.attn      = None # Attention window
+        self.word_dict = SymbolTable() # Symbol table for dictionary
+        # Define input layers
+        self.sentences        = None
+        self.sentence_lengths = None
+        self.train_marginals  = None
+        self.keep_prob        = None
+        self.seed             = seed
+        # Super constructor
+        super(LSTMBase, self).__init__(
+            n_threads=n_threads, save_file=save_file, name=name
+        )
+
+    def _preprocess_data(self, candidates, extend):
+        raise NotImplementedError()
+
+    def _make_tensor(self, x):
+        """Construct input tensor with padding"""
+        batch_size = len(x)
+        x_batch    = np.zeros((batch_size, self.mx_len), dtype=np.int32)
+        len_batch  = np.zeros(batch_size, dtype=np.int32)
+        for j, token_ids in enumerate(x):
+            t               = min(len(token_ids), self.mx_len)
+            x_batch[j, 0:t] = token_ids[0:t]
+            len_batch[j]    = t
+        return x_batch, len_batch
+
+    def _embedding_init(self, s):
+        return tf.random_normal((self.n_v-1, self.dim), stddev=SD, seed=s)
+
+    def _build(self):
+        """Get feed forward step, loss function, and optimizer for LSTM"""
+        # Define input layers
+        self.sentences        = tf.placeholder(tf.int32, [None, None])
+        self.sentence_lengths = tf.placeholder(tf.int32, [None])
+        self.train_marginals  = tf.placeholder(tf.float32, [None])
+        self.keep_prob        = tf.placeholder(tf.float32)
+        # Seeds
+        s = self.seed
+        s1, s2, s3, s4 = [None] * 4 if s is None else [s+i for i in range(4)]
+        # Embedding layer
+        emb_var   = tf.Variable(self._embedding_init(s1))
+        embedding = tf.concat([tf.zeros([1, self.dim]), emb_var], axis=0)
+        inputs    = tf.nn.embedding_lookup(embedding, self.sentences)
+        # Build RNN graph
+        batch_size = tf.shape(self.sentences)[0]
+        rand_name  = "LSTM_{0}".format(random.randint(0, 1e12)) # Obscene hack
+        init = tf.contrib.layers.xavier_initializer(seed=s2)
```

Comment:
Get deep bro. This is pretty standard for deep nets, probably ok to leave without comment.

http://jmlr.org/proceedings/papers/v9/glorot10a/glorot10a.pdf